# <font color='blue'>Business Analytics</font>

# <font color='blue'>Capítulo 7 - Marketing Analytics</font>
# <font color='blue'>Projeto 4</font>
## <font color='blue'>Otimização de Preços e Mix de Produtos</font>

In [1]:
# Versão da Linguagem Python
from platform import python_version
print('Versão da Linguagem Python Usada Neste Jupyter Notebook:', python_version())

Versão da Linguagem Python Usada Neste Jupyter Notebook: 3.9.13


![title](imagens/projeto4.png)

## Marketing Analytics

Marketing Analytics compreende os processos e tecnologias que permitem aos profissionais de Marketing avaliar o sucesso de suas iniciativas. 

Isso é feito medindo o desempenho das campanhas de Marketing, coletando os dados e analisando os resultados. Marketing Analytics utiliza métricas importantes de negócios, como ROI (Retorno Sobre o Investimento), Atribuição de Marketing e Eficácia Geral do Marketing. Em outras palavras, o Marketing Analytics mostra se os programas de Marketing estão sendo efetivos ou não.

Marketing Analytics reúne dados de todos os canais de marketing e os consolida em uma visão de marketing comum. A partir dessa visão comum, você pode extrair resultados analíticos que podem fornecer assistência inestimável para impulsionar os esforços de marketing.

## Definindo o Problema

Leia o manual em pdf no Capítulo 7.

## Instalando e Carregando os Pacotes

In [2]:
# Para atualizar um pacote, execute o comando abaixo no terminal ou prompt de comando:
# pip install -U nome_pacote

# Para instalar a versão exata de um pacote, execute o comando abaixo no terminal ou prompt de comando:
# !pip install nome_pacote==versão_desejada

# Depois de instalar ou atualizar o pacote, reinicie o jupyter notebook.

# Instala o pacote watermark. 
# Esse pacote é usado para gravar as versões de outros pacotes usados neste jupyter notebook.
!pip install -q -U watermark

In [3]:
# Instala o PuLP
# https://pypi.org/project/PuLP/
# https://coin-or.github.io/pulp/
!pip install -q pulp

In [4]:
# Imports
from pulp import *

In [5]:
# Versões dos pacotes usados neste jupyter notebook
%reload_ext watermark
%watermark -a "Versions" --iversions

Author: Versions

pulp   : 2.7.0
re     : 2.2.1
logging: 0.5.1.2
json   : 2.0.9
ctypes : 1.1.0
sys    : 3.9.13 (main, Aug 25 2022, 18:29:29) 
[Clang 12.0.0 ]



## Criando o Modelo Matemático Para a Otimização

### Parâmetros

- Ai = Número máximo de smartphones modelo tipo i para vender este mês, onde i pertence ao conjunto {Lua1, Lua2}
- Bi = Preço de venda de smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Ci = Preço de custo das peças componentes para smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Di = Custo de mão de obra de montagem por hora de smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Ei = Custo de mão de obra de teste por hora de smartphones modelo tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- F = Número máximo de horas de trabalho de montagem
- G = Número máximo de horas de trabalho de teste
- Hf,i = Horas de montagem necessárias para construir cada modelo de smartphone tipo i, onde i pertence ao conjunto {Lua1, Lua2}
- Hg,i = Horas de teste necessárias para testar cada modelo de smartphone tipo i, onde i pertence ao conjunto {Lua1, Lua2}

### Variável de Decisão

- Xi = Número de smartphones modelo tipo i a produzir este mês, onde i pertence ao conjunto {Lua1, Lua2}

### Função Objetivo

![title](imagens/formula.png)

### Restrições

- O número de smartphones modelo tipo i a serem produzidos não pode ser negativo, ou seja, Xi >= 0, onde i pertence ao conjunto {Lua1, Lua2}.


- O número total de horas de montagem não pode ser maior que o número máximo de horas de mão de obra de montagem disponíveis.


- O número total de horas de teste não pode ser maior do que o máximo de horas de mão de obra de teste disponíveis.


- O número de smartphones modelo tipo i a serem produzidos não pode ser maior do que o número máximo de smartphones modelo tipo i a serem vendidos neste mês, onde i pertence ao conjunto {Lua1, Lua2}.

## Implementando o Modelo Matemático

### Organizando os Parâmetros

In [6]:
# Número máximo de smartphones para vender este mês
A_Lua1 = 600
A_Lua2 = 1200

In [7]:
# Preço de venda de smartphones 
B_Lua1 = 300 
B_Lua2 = 450 

In [8]:
# Preço de custo das peças componentes para smartphones
C_Lua1 = 150
C_Lua2 = 225

In [9]:
# Custo de mão de obra de montagem por hora de smartphones
D_Lua1 = 11
D_Lua2 = 11

In [10]:
# Custo de mão de obra de teste por hora de smartphones 
E_Lua1 = 15 
E_Lua2 = 15 

In [11]:
# Número máximo de horas de trabalho de montagem
F = 10000 

In [12]:
# Número máximo de horas de trabalho de teste
G = 3000 

In [13]:
# Horas de montagem necessárias para construir cada modelo de smartphone
Hfi_Lua1 = 5
Hfi_Lua2 = 6

In [14]:
# Horas de teste necessárias para testar cada modelo de smartphone
Hgi_Lua1 = 1 
Hgi_Lua2 = 2

### Criação da Variável Para o Problema de Otimização

In [15]:
help(LpProblem)

Help on class LpProblem in module pulp.pulp:

class LpProblem(builtins.object)
 |  LpProblem(name='NoName', sense=1)
 |  
 |  An LP Problem
 |  
 |  Methods defined here:
 |  
 |  __getstate__(self)
 |  
 |  __iadd__(self, other)
 |  
 |  __init__(self, name='NoName', sense=1)
 |      Creates an LP Problem
 |      
 |      This function creates a new LP Problem  with the specified associated parameters
 |      
 |      :param name: name of the problem used in the output .lp file
 |      :param sense: of the LP problem objective.                  Either :data:`~pulp.const.LpMinimize` (default)                 or :data:`~pulp.const.LpMaximize`.
 |      :return: An LP Problem
 |  
 |  __repr__(self)
 |      Return repr(self).
 |  
 |  __setstate__(self, state)
 |  
 |  add(self, constraint, name=None)
 |  
 |  addConstraint(self, constraint, name=None)
 |  
 |  addVariable(self, variable)
 |      Adds a variable to the problem before a constraint is added
 |      
 |      @param variable:

In [16]:
# Variável para o problema
problema = LpProblem("MixProdutos", LpMaximize)

In [17]:
# Visualiza
problema

MixProdutos:
MAXIMIZE
None
VARIABLES

### Definindo a Variável de Decisão de Cada Modelo de Smartphone

In [18]:
help(LpVariable)

Help on class LpVariable in module pulp.pulp:

class LpVariable(LpElement)
 |  LpVariable(name, lowBound=None, upBound=None, cat='Continuous', e=None)
 |  
 |  This class models an LP Variable with the specified associated parameters
 |  
 |  :param name: The name of the variable used in the output .lp file
 |  :param lowBound: The lower bound on this variable's range.
 |      Default is negative infinity
 |  :param upBound: The upper bound on this variable's range.
 |      Default is positive infinity
 |  :param cat: The category this variable is in, Integer, Binary or
 |      Continuous(default)
 |  :param e: Used for column based modelling: relates to the variable's
 |      existence in the objective function and constraints
 |  
 |  Method resolution order:
 |      LpVariable
 |      LpElement
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __init__(self, name, lowBound=None, upBound=None, cat='Continuous', e=None)
 |      Initialize self.  See help(type(self)) for 

In [19]:
# Define as variáveis
x_Lua1 = LpVariable("Unidades Lua1", 0, None, LpInteger)
x_Lua2 = LpVariable("Unidades Lua2", 0, None, LpInteger)

In [20]:
# Print
print(x_Lua1)
print(x_Lua2)

Unidades_Lua1
Unidades_Lua2


### Implementando a Função Objetivo

![title](imagens/formula.png)

In [21]:
# Faturamento
faturamento = (x_Lua1 * B_Lua1) + (x_Lua2 * B_Lua2)
faturamento

300*Unidades_Lua1 + 450*Unidades_Lua2 + 0

In [22]:
# Custo de Montagem
custo_montagem = (x_Lua1 * Hfi_Lua1 * D_Lua1) + (x_Lua2 * Hfi_Lua2 * D_Lua2)
custo_montagem

55*Unidades_Lua1 + 66*Unidades_Lua2 + 0

In [23]:
# Custo de Teste
custo_teste = (x_Lua1 * Hgi_Lua1 * E_Lua1) + (x_Lua2 * Hgi_Lua2 * E_Lua2)
custo_teste

15*Unidades_Lua1 + 30*Unidades_Lua2 + 0

In [24]:
# Custo de Componentes
custo_componentes = (x_Lua1 * C_Lua1) + (x_Lua2 * C_Lua2) 
custo_componentes

150*Unidades_Lua1 + 225*Unidades_Lua2 + 0

In [25]:
# Lucro = Faturamento - Custo de Montagem - Custo de Teste - Custo de Componentes
problema += faturamento - custo_montagem - custo_teste - custo_componentes
problema

MixProdutos:
MAXIMIZE
80*Unidades_Lua1 + 129*Unidades_Lua2 + 0
VARIABLES
0 <= Unidades_Lua1 Integer
0 <= Unidades_Lua2 Integer

### Restrições

In [26]:
# Número máximo de horas de montagem
problema += (x_Lua1 * Hfi_Lua1) + (x_Lua2 * Hfi_Lua2) <= F,"Número Máximo de Horas de Montagem"

In [27]:
# Número máximo de horas de teste
problema += (x_Lua1 * Hgi_Lua1) + (x_Lua2 * Hgi_Lua2) <= G,"Número Máximo de Horas de Teste"

In [28]:
# Produção menor ou igual a demanda pelo modelo Lua1
problema += x_Lua1 <= A_Lua1,"Produção menor ou igual a demanda pelo modelo Lua1"

In [29]:
# Produção menor ou igual a demanda pelo modelo Lua2
problema += x_Lua2 <= A_Lua2,"Produção menor ou igual a demanda pelo modelo Lua2"

In [30]:
# Problema final
problema

MixProdutos:
MAXIMIZE
80*Unidades_Lua1 + 129*Unidades_Lua2 + 0
SUBJECT TO
Número_Máximo_de_Horas_de_Montagem: 5 Unidades_Lua1 + 6 Unidades_Lua2 <= 10000

Número_Máximo_de_Horas_de_Teste: Unidades_Lua1 + 2 Unidades_Lua2 <= 3000

Produção_menor_ou_igual_a_demanda_pelo_modelo_Lua1: Unidades_Lua1 <= 600

Produção_menor_ou_igual_a_demanda_pelo_modelo_Lua2: Unidades_Lua2 <= 1200

VARIABLES
0 <= Unidades_Lua1 Integer
0 <= Unidades_Lua2 Integer

### Resolvendo o Problema de Otimização

In [31]:
# Otimização
problema.solve()

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/pedropedrosa/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/l1/1dxjhd1d47qbqjtdrr3ms2y80000gn/T/b40985b5ec7f4c2a9ad265fd5b5bbd98-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/l1/1dxjhd1d47qbqjtdrr3ms2y80000gn/T/b40985b5ec7f4c2a9ad265fd5b5bbd98-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 9 COLUMNS
At line 22 RHS
At line 27 BOUNDS
At line 30 ENDATA
Problem MODEL has 4 rows, 2 columns and 6 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 199600 - 0.00 seconds
Cgl0004I processed model has 1 rows, 2 columns (2 integer (0 of which binary)) and 2 elements
Cutoff increment increased from 1e-05 to 0.9999
Cbc0012I Integer solution of -199600 found by DiveCoefficient after 0 iterations and 0 nodes (0.00 seconds)
Cbc0001I

1

In [32]:
# Lucro Maximizado
print("Lucro Maximizado:", value(problema.objective))

Lucro Maximizado: 199600.0


In [33]:
# Número de Unidades do Modelo Lua 1 a Produzir
print("Número de Unidades do Modelo Lua 1 a Produzir:", problema.variables()[0].varValue)

Número de Unidades do Modelo Lua 1 a Produzir: 560.0


In [34]:
# Número de Unidades do Modelo Lua 1 a Produzir
print("Número de Unidades do Modelo Lua 2 a Produzir:", problema.variables()[1].varValue)

Número de Unidades do Modelo Lua 2 a Produzir: 1200.0


##  Conclusão

A empresa Lua Smart Tech deve produzir 560 unidades do tipo de smartphone Lua1 e 1200 do tipo Lua2 para atingir o lucro máximo de R$199.600.

# Fim